### ISSUE TITLE
0.8.7b10 colreorder - server side issues

### ISSUE THREAD

**Issue by johndegey** [vedmack/yadcf#151]
Hello,

I've tested the beta10...

1) i remove the inclusion of the colreoder.js file and the capital R from the datatable DOM parameter
 -> if the yadcf are in the footer : no problem, my select filters have the good options and filters the good column
-> if the adcf filters are in the header, my select filter haven't got the good option value? There is a shift (i have the hidden column and the first with no filter at all). 

2) i do not remove the inclusion of the colreorder.js file but i only remove the capital R from the datable DOM parameter
 - i have an error in firebug and filtering does not work at all
TypeError: this.context[0]._colReorder is undefined

3) my previous issue is still present
 https://github.com/vedmack/yadcf/issues/150 

Thanks for your work, hope this will help you

**Comment by vedmack** [comment]
Please provide a working samples that reproduces the issues specified above. And IMO those issues are not related to colreorder because its issues arise only after doing the columns reordering which causing column index to be "confused".

**Comment by luivis07** [comment]
Hi Daniel, I'm not sure if this is the issue that OP is referring to but basically what is happening is that there is an additional filter container being created. If you look at the link I'm including here: http://tempdatatables.azurewebsites.net/ (I had to use this instead of a fiddle because it only happens server side). If you drag "Header 4" to position one and then reload the page there will be an additional filter in "name (1)" column. 
For debugging purposes I've added two textbox at the top left of the page, the first one indicates the columns that had a filter originally, and the second textbox shows the "column_number" property from the "filter" object passed into the yadcf.init 
The table is using ColReorder and StateSave = true.
Let me know if you have any questions or if I can help in anyway. Thanks.

**Comment by johndegey** [comment]
Hello,

this one is with filter in the footer
http://yadcf.johndegey.org/listing.php

this one is with filter in the header
http://yadcf.johndegey.org/listing2.php

if you hide field1 and refresh, listing2 will have bad filter values for select

if you switch field2 and field3, next use field3's select filter, you'll see the problem

Thanks again

**Comment by luivis07** [comment]
@vedmack by the way the filters get added correctly when using version 0.8.6 maybe that bug was introduced with the fix for colReorder applying the filter to the correct column?

**Comment by vedmack** [comment]
@johndegey  and @luivis07 , go get the 0.8.7.beta.11 , let me know how it works for you (this specific fix is for ColReorder + stateSave: true

**Comment by johndegey** [comment]
Hello Daniel,

The shift problem when i hide the first column seems to be fix.
Select options in the header or the footer have the good values.

The problem is still there if you apply a filter to col3, then shift col3 and col4 and refresh the page.
Here two new link with the b11.
http://yadcf.johndegey.org/listing3.php

http://yadcf.johndegey.org/listing4.php

John

**Comment by vedmack** [comment]
@johndegey , please use this yadcf link file in your examples , so will be able to debug / fix it better and assign single meaningull value/filter_default_label into each select filter, e.g in field2 column place single value "field2_val" and assign it with label "field2_label" that will ease the debugging too

**Comment by johndegey** [comment]
@vedmack i've made the modifications, is it good for u ?

**Comment by vedmack** [comment]
Nope, please assign single meaningull values to each filter **and**  a meaningful `filter_default_label` to each filter. thanks

**Comment by johndegey** [comment]
ok, all select have just one option value, and i set the filter_default_value for each filter, including the text ones

**Comment by vedmack** [comment]
I did fixed some, but... I noticed that you send the yadcf_data_X based on the reordering of the columns, the thing is that I expect that X means the original X before the re ordering of columns and yadcf will know to redirect it later on to the new position of X in the table, so my question is: can you fixate that data that belong to the original column number X? (in other words, yadcf_data_2 should hold the same values of the original column 2 no matter its current position)  <-- can it be done on your side? While its not a big deal to add additional code on yadcf to take into consideration the new position of the columns when getting the yadcf_data_X I pretty sure that while I debugged your page I noticed that you are sending wrong data in those yadcf_data_X / yadcf_data__Y etc'. so I think it will be better the send those yadcf_data_X fixed to the original location of the columns.

Regarding the ctrl F5 and next/prev isn't its some datatables inconsistency and not related to yadcf?

**Comment by johndegey** [comment]
I only have the column number send by DT. But i don't even have to use it at all because i set the "name" and "data" in the column parameter of DT. (i have to test but normally, i could remove the "name" and just keep the "data")
This allows me to send back rows from my sql request with no post treatment for re-ordering according to actual position in the table. 
I use the same server code for multiple templates, so i can't know the original column order.

Maybe i can use some pre-treatment on the client side to add original column number to the xhr request, but i'm not sure this is the good way. Not that i don't want to try, but i really think that the plugin should do the re-ordering.
Based on actual number ? 
Based on the column's name parameter ?. 
Personally, I would prefer the name attribute.

Here is the portion of code for setting the yadcf_data_x

	$this->aColumnsFilterValues = array(
		'field2' => array(array('value' => 'field2_val', 'label' => 'field2_label')),
		'field3' => array(array('value' => 'field3_val', 'label' => 'field3_label')),
		'field4' => array(array('value' => 'field4_val', 'label' => 'field4_label')),
		'field5' => array(array('value' => 'field5_val', 'label' => 'field5_label'))
	);
	foreach ($_POST['columns'] as $colNumber => $colInfos)
	{
		if (array_key_exists($colInfos['name'],$this->aColumnsFilterValues) !== false)
		{
			$output['yadcf_data_'.$colNumber] = $this->aColumnsFilterValues[$colInfos['name']];
		}
	}
I don't see why sometimes it should send data to wrong column number.


For the CRTL F5, i don't know... i haven't do debugging for this one yet.

Thx

**Comment by vedmack** [comment]
Regarding the columns[1][data] field2 (in F5)  and columns[1][data] field3 (Paginate to next page) its really wrong IMO , I think that its must be consistent in both scenarios, so its either a your local server bug or a bug in datatables, I mean... why do you pass different values on F5 and on paginate?

**Comment by johndegey** [comment]
I 've posted in the DT forum

http://datatables.net/forums/discussion/25858/server-side-new-api-statesave-colreorder-and-column-s-order-in-the-xhr-request#latest

Wait & see...

**Comment by luivis07** [comment]
@vedmack the fix you applied for stateSave = true, what exactly is it doing? because in the example here : http://tempdatatables.azurewebsites.net/, I'm already getting the ColReorder from localstorage and comparing it to the original order and setting the "column_number" to the correct value. 
I think maybe in the fix you are re-applying this logic? You can inspect the element and scroll down the js function GetNewIndex(columnNum, newOrder)

**Comment by vedmack** [comment]
@luivis07 , please disable your logic on that you have mentioned above (it messes the original column_number and eventually yadcf too) , and also use [this yadcf link](https://dl.dropboxusercontent.com/u/55621125/yadcf/war14/resources/js/jquery.dataTables.yadcf.js) for better debug

**Comment by luivis07** [comment]
@vedmack after disabling the logic the filter are being added to the right columns (the only thing left now is #138). I do know about the 'yadcf_data_xxx' just didn't bother setting that up for the demo.

**Comment by vedmack** [comment]
@luivis07 are you using statesave too? And can you confirm it works for F5 and pagination ?

**Comment by luivis07** [comment]
@vedmack yes I'm using stateSave, pagination work, by F5  you mean refresh right? if so that works too.

**Comment by vedmack** [comment]
Great news @luivis07 , so... @johndegey ,  maybe you compare your code with @luivis07 ?

**Comment by johndegey** [comment]
@luivis07 : is this the good link for your test case ?
http://tempdatatables.azurewebsites.net/

**Comment by luivis07** [comment]
@johndegey yes that should be working now.

**Comment by johndegey** [comment]
@vedmack there is now way i can add the original index to the xhr query right now
see here http://datatables.net/forums/discussion/25952/how-to-get-dt-api-in-ajax-data-function-with-server-side#latest

Is there a way that you add the possibility to use column's name like yadcf_data_name ?

@luivis07 your test case is not complete
- the filtering is not working
- you never send back any yadcf_data_xxx
- i can just confirm that reloading the page or using the paging system has the same behavior than my test case 
=> is it possible for you to make it work completely on http://tempdatatables.azurewebsites.net/

John

**Comment by vedmack** [comment]
@johndegey , I think I understood what I was missing all this time, if you place a breakpoint on `$(document).off('xhr.dt', oTable.selector).on('xhr.dt', oTable.selector, function (e, settings, json) {` code and debug the F5 flow and pagination flow you will notice that the datatable is in different states, while in F5 the columns are in their original location, in the pagination flow the columns are already in their reordered location, (you can notice this by looking at the frozen browser view... well no I need to rethink how I should approach this case. Since I know now why I get what I get I hope it wont take much time. regarding the `yadcf_data_name` I guess you can open a new issue for that...

**Comment by luivis07** [comment]
@johndegey to be honest, it would take some time for me to set up the filtering and pagination server side since I just put that page up for 1 specific reason. However, I don't see how adding the filtering will help in reproducing the bug? about the yadcf_data_xxx, this goes back to the same thing I just didn't bother setting it up.

**Comment by johndegey** [comment]
@luivis07 : without settings a correct yadcf_data_x response, i can't compare with my test case... but i think it is not necessary anymore ;)

@vedmack thanks for your work

**Comment by vedmack** [comment]
@johndegey , you can test your link now: http://yadcf.johndegey.org/listing4.php , I've tuned some code in yadcf (0.8.7.beta.14 on my dropbox) and it seems to work as expected

**Comment by johndegey** [comment]
great... everything works fine, did not find any problem when in the header position
can you do the same when the filters are in the footer
http://yadcf.johndegey.org/listing3.php

i will try again the combination of FixedHeader and the beta14 (did not work previously)

Thx

**Comment by vedmack** [comment]
Should work now with beta15. I still need to check how show and hide works when filters in footer

**Comment by vedmack** [comment]
Looks like I'm done with this issue, should work for all combinations of header/footer/hidden columns, regarding the FixedHeader - its another feature that I will need to add support and not related to the current issue...

**Comment by johndegey** [comment]
I've found a problem when in the footer (works in the header)
All column in order.
Select 'field_label2' in the field2_select
Put field2 after field3 and then hide field3 => problem with paging and F5


(i will open another case for FixedHeader when this one will be closed)

**Comment by vedmack** [comment]
@johndegey , 0.8.7.beta.17 is out and it looks like this issue came to an end :) , thanks for all your great test cases.
